src/componententity.cpp

```diff
@@ -375,257 +282,164 @@ void ComponentEntity::doAddComponent(const ComponentPtr &c)
 }
 
-ComponentPtr ComponentEntity::ComponentEntityImpl::takeComponent(const std::string &name, bool searchEncapsulated)
+ComponentPtr ComponentEntity::takeComponent(const std::string &name, bool searchEncapsulated)
 {
     ComponentPtr foundComponent = nullptr;
-    if (containsComponentInThis(name)) {
-        foundComponent = takeComponentInThis(name);
+    auto result = mPimpl->findComponent(name);
+    if (result != mPimpl->mComponents.end()) {
+        foundComponent = *result;
+        mPimpl->mComponents.erase(result);
     } else if (searchEncapsulated) {
-        for (size_t i = 0; i < componentCount(); ++i) {
+        for (size_t i = 0; i < componentCount() && !foundComponent; ++i) {
             foundComponent = getComponent(i)->takeComponent(name, searchEncapsulated);
-            if (foundComponent) {
-                break;
-            }
         }
     }
 
     return foundComponent;
 }
 
-ComponentPtr ComponentEntity::takeComponent(size_t index)
-{
-    return mPimpl->takeComponent(index);
-}
-
-ComponentPtr ComponentEntity::takeComponent(const std::string &name, bool searchEncapsulated)
-{
-    return mPimpl->takeComponent(name, searchEncapsulated);
-}
-
-void ComponentEntity::replaceComponent(size_t index, const ComponentPtr &c)
-{
-    mPimpl->replaceComponent(index, c);
-}
-
-void ComponentEntity::replaceComponent(const std::string &name, const ComponentPtr &component, bool searchEncapsulated)
+bool ComponentEntity::replaceComponent(size_t index, const ComponentPtr &c)
 {
-    mPimpl->replaceComponent(name, component, searchEncapsulated);
-}
-
-void ComponentEntity::ComponentEntityImpl::replaceComponent(size_t index, const ComponentPtr &c)
-{
-    removeComponent(index);
-    mComponents.insert(mComponents.begin() + index, c);
-}
-
-void ComponentEntity::ComponentEntityImpl::replaceComponent(const std::string &name, const ComponentPtr &component, bool searchEncapsulated)
-{
-    if (containsComponentInThis(name)) {
-        replaceComponentInThis(name, component);
-    } else if (searchEncapsulated) {
-        for (size_t i = 0; i < componentCount(); ++i) {
-            if (getComponent(i)->containsComponent(name, searchEncapsulated)) {
-                getComponent(i)->replaceComponent(name, component, searchEncapsulated);
-                break;
-            }
-        }
+    bool status = false;
+    if (index < mPimpl->mComponents.size()) {
+        removeComponent(index);
+        mPimpl->mComponents.insert(mPimpl->mComponents.begin() + index, c);
+        status = true;
     }
-}
-
-// Begin private component methods.
-ComponentPtr ComponentEntity::ComponentEntityImpl::takeComponentInThis(const std::string &name)
-{
-    return takeComponent(findComponent(name) - mComponents.begin());
-}
 
-void ComponentEntity::ComponentEntityImpl::replaceComponentInThis(const std::string &name, const ComponentPtr &component)
-{
-    replaceComponent(findComponent(name) - mComponents.begin(), component);
+    return status;
 }
```

Comment:
Same here, take advantage of the fact that `removeComponent()` now returns a boolean.